### PR TITLE
accounts/abi/bind, internal/ethapi: binary search gas estimation

### PR DESCRIPTION
**Pull request by karalabe**
Gas estimation currently mostly works, but can underestimate for more funky refunds. This is because various ops (e.g. CALL) need more gas to run than they actually consume (e.g. 2300 stipend that is refunded if not used). With more intricate contract interplays, it becomes almost impossible to return a proper value to the user.

This PR swaps out the simplistic gas estimation to a binary search approach, honing in on the correct gas use. This does mean that gas estimation needs to rerun the transaction log(max-price) times to measure whether it fails or not, but it's a price paid by the transaction issuer, and it should be worth it to support proper estimates.

Fixes: #3282 #2970 #1590 

---

Verified in both the simulator tests as well as on Ropsten:

```
contract FunkyGasPattern {
	string public field;

	function SetField(string value) {
		// This check will screw gas estimation! Good, good!
		if (msg.gas < 100000) {
			throw;
		}
		field = value;
	}
}
```

Deployed at [0xf347ee4bF34a1a4e5c7a11694669602b5324Fc56](https://testnet.etherscan.io/address/0xf347ee4bF34a1a4e5c7a11694669602b5324Fc56)

Calling SetField is estimated as

```
> eth.estimateGas({from: eth.accounts[0], to: "0xf347ee4bF34a1a4e5c7a11694669602b5324Fc56", data: "0x23fcf32a0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000564756e6e6f000000000000000000000000000000000000000000000000000000"})
122418
```

### Running with this estimated gas cost:

```
> eth.sendTransaction({from: eth.accounts[0], to: "0xf347ee4bF34a1a4e5c7a11694669602b5324Fc56", gas: "122418", data: "0x23fcf32a0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000564756e6e6f000000000000000000000000000000000000000000000000000000"})
"0x31912e1743c5db3a9af4739c937fa1b595277e7d445b12c977dcecf2230430da"
```

Results in successful transaction [0x31912e1743c5db3a9af4739c937fa1b595277e7d445b12c977dcecf2230430da](https://testnet.etherscan.io/tx/0x31912e1743c5db3a9af4739c937fa1b595277e7d445b12c977dcecf2230430da)

**Note, actual gas usage is 42918, way below the estimated 122418!**

### Running with 1 gas below the estimated cost:

```
> eth.sendTransaction({from: eth.accounts[0], to: "0xf347ee4bF34a1a4e5c7a11694669602b5324Fc56", gas: "122417", data: "0x23fcf32a0000000000000000000000000000000000000000000000000000000000000020000000000000000000000000000000000000000000000000000000000000000564756e6e6f000000000000000000000000000000000000000000000000000000"})
"0xffb8ffbf722de204781ba9f94c7c4065edc03de7bbfb5583b859b782ee815718"
```

Results in a failed transaction [0xffb8ffbf722de204781ba9f94c7c4065edc03de7bbfb5583b859b782ee815718](https://testnet.etherscan.io/tx/0xffb8ffbf722de204781ba9f94c7c4065edc03de7bbfb5583b859b782ee815718)

**Note, even though the end gas usage would have been 42918, the contract goes OOG with only 122417 allowance!**

### CODE REVIEW

**Comment by holiman** [comment]
Hm. That's not quite accurate ... It doesn't go out of gas, you're triggering a bad jump put there by Solidity

**Comment by holiman** [comment]
But I agree that this approach is probably better in general.

**Comment by karalabe** [comment]
@holiman Well yeah, but it simulates a contract requiring more gas to run than actually consumed, which is the corner case estimated badly.

File: internal/ethapi/api.go [hunk 1]
```diff
@@ -559,8 +559,34 @@ func (s *PublicBlockChainAPI) Call(ctx context.Context, args CallArgs, blockNr r
...
+
+		// If the transaction became invalid or used all the gas (failed), raise the gas limit
+		if err != nil || gas.Cmp((*big.Int)(&args.Gas)) == 0 {
+			lo = mid
+			continue
+		}
+		// Otherwise assume the transaction succeeded, lower the gas limit
+		hi = mid
+	}
+	return (*hexutil.Big)(big.NewInt(int64(hi))), nil
```
Comment (karalabe) [inline internal/ethapi/api.go]: new(big.Int).SetUint64()

**Comment by karalabe** [comment]
@fjl PTAL

**Pull request merged**